File: python/app/export/svg_exporter.py
"""SVG export (DrawingBot V3 compatible)."""

from __future__ import annotations

from typing import Dict, List

from app.core.drawing_area import DrawingArea
from app.core.drawing_pen import DrawingPen
from app.export.path_export import prepare_pen_paths


def export_svg(
    filepath: str,
    geometries,
    drawing_area: DrawingArea,
    active_pens: List[DrawingPen],
    path_opt: Dict,
) -> None:
    w_mm = drawing_area.width_mm
    h_mm = drawing_area.height_mm
    pen_paths = prepare_pen_paths(geometries, path_opt)

    lines = [
        '<?xml version="1.0" encoding="UTF-8"?>',
        f'<svg xmlns="http://www.w3.org/2000/svg" '
        f'xmlns:inkscape="http://www.inkscape.org/namespaces/inkscape" '
        f'width="{w_mm}mm" height="{h_mm}mm" '
        f'viewBox="0 0 {w_mm} {h_mm}">',
    ]

    for pi in sorted(pen_paths.keys()):
        pen = active_pens[pi % len(active_pens)]
        stroke_w = pen.stroke_width * 0.264583
        lines.append(f'  <g id="pen_{pi}" inkscape:label="{pen.name}">')
        for path in pen_paths[pi]:
            d_parts = [f"M{path[0][0]:.3f},{path[0][1]:.3f}"]
            for x, y in path[1:]:
                d_parts.append(f"L{x:.3f},{y:.3f}")
            d = " ".join(d_parts)
            lines.append(
                f'    <path d="{d}" fill="none" stroke="{pen.color_hex}" '
                f'stroke-width="{stroke_w:.4f}" stroke-linecap="round" '
                f'stroke-linejoin="round"/>'
            )
        lines.append("  </g>")

    lines.append("</svg>")
    with open(filepath, "w", encoding="utf-8") as f:
        f.write("\n".join(lines))

File: cpp_backup/export/svg_exporter.cpp
#include "export/svg_exporter.h"
#include <QFile>
#include <QTextStream>
#include <QMap>

bool SVGExporter::exportSVG(const QString& filepath,
                            const QVector<DrawingGeometry>& geometries,
                            const DrawingAreaConfig& drawingArea,
                            int imageWidth,
                            int imageHeight,
                            const QVector<QColor>& penColors,
                            double penWidthMm) {
    if (geometries.isEmpty()) return false;

    QFile file(filepath);
    if (!file.open(QIODevice::WriteOnly | QIODevice::Text)) {
        return false;
    }

    QTextStream out(&file);

    // Write SVG header
    out << "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n";
    out << QString("<svg xmlns=\"http://www.w3.org/2000/svg\" "
                   "xmlns:inkscape=\"http://www.inkscape.org/namespaces/inkscape\" "
                   "width=\"%1mm\" height=\"%2mm\" "
                   "viewBox=\"0 0 %1 %2\">\n")
               .arg(drawingArea.width_mm).arg(drawingArea.height_mm);

    // Group paths by pen index
    QMap<int, QVector<Path>> penPaths;
    for (const auto& geom : geometries) {
        Path mmPath = pixel_path_to_mm(geom.path, imageWidth, imageHeight, drawingArea);
        if (!mmPath.empty()) {
            penPaths[geom.penIndex].append(mmPath);
        }
    }

    // Write groups
    for (auto it = penPaths.begin(); it != penPaths.end(); ++it) {
        int penIdx = it.key();
        QColor color = (penIdx >= 0 && penIdx < penColors.size()) ? penColors[penIdx] : QColor(0, 0, 0);
        
        out << QString("  <g id=\"pen_%1\" inkscape:label=\"Pen %1\">\n").arg(penIdx);
        
        for (const auto& path : it.value()) {
            if (path.size() < 2) continue;
            
            QString d = QString("M%1,%2").arg(path[0].first, 0, 'f', 3).arg(path[0].second, 0, 'f', 3);
            for (std::size_t i = 1; i < path.size(); ++i) {
                d += QString(" L%1,%2").arg(path[i].first, 0, 'f', 3).arg(path[i].second, 0, 'f', 3);
            }
            
            out << QString("    <path d=\"%1\" fill=\"none\" stroke=\"%2\" "
                           "stroke-width=\"%3\" stroke-linecap=\"round\" "
                           "stroke-linejoin=\"round\"/>\n")
                       .arg(d).arg(color.name()).arg(penWidthMm, 0, 'f', 4);
        }
        
        out << "  </g>\n";
    }

    out << "</svg>\n";
    return true;
}
